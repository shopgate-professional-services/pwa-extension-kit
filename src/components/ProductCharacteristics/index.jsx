import React from 'react';
import BaseProductCharacteristics from '@shopgate/pwa-common/components/ProductCharacteristics';
import withProductContext from '../../connectors/withProductContext';

export default withProductContext(({ productContext, render }) => {
  const newProps = { ...productContext, render };
  return <BaseProductCharacteristics {...newProps} />;
});
