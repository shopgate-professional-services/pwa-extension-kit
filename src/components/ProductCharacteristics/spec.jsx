import React from 'react';
import { mount } from 'enzyme';
import Conditioner from '@shopgate/pwa-core/classes/Conditioner';
import ProductCharacteristics from './index';

const mockedConditioner = new Conditioner();
const mockedMapStateToPropsResult = {
  variants: {
    products: [{
      id: '32050590001000',
      characteristics: { 1: '1', 2: '11' },
    }, {
      id: '32050590000950',
      characteristics: { 1: '2', 2: '12' },
    }],
    characteristics: [
      {
        id: '1',
        label: 'Color',
        values: [
          { id: '1', label: 'BLACK-WHITE (9000)' },
          { id: '2', label: 'RED-WHITE (1000)' }],
      },
      {
        id: '2',
        label: 'Size',
        values: [
          { id: '11', label: '7' }, { id: '12', label: '7.5' },
        ],
      }],
  },
};
jest.mock('react-redux', () => ({
  connect: mapStateToProps => Component => props => (
    <Component
      variants={mockedMapStateToPropsResult.variants}
      {...props}
    />
  ),
}));

jest.mock('@shopgate/pwa-common/context', () => ({
  // eslint-disable-next-line react/prop-types
  Theme: ({ children, ...otherProps }) => {
    const Child = children;

    const props = {
      AppBar: () => null,
      Drawer: () => null,
      View: () => null,
      contexts: {
        ProductContext: {
          // eslint-disable-next-line react/prop-types
          Consumer: ({ children: ContextChildren, ...contextProps }) => (
            <ContextChildren
              options={{}}
              productId="123"
              variantId="123-45"
              conditioner={mockedConditioner}
              {...contextProps}
            />
          ),
        },
      },
      ...otherProps,
    };

    return (
      <Child {...props} />
    );
  },
}));

describe('components/ProductCharacteristics', () => {
  it('should render', () => {
    const component = mount((
      <ProductCharacteristics render={jest.fn()} />
    ));

    expect(component).toMatchSnapshot();
  });

  it('should call render prop', () => {
    const renderer = jest.fn();

    mount((
      <ProductCharacteristics render={renderer} />
    ));

    const expected1 = {
      charRef: { current: null },
      disabled: false,
      highlight: false,
      id: '1',
      key: '1',
      label: 'Color',
      selected: null,
      values: [
        { id: '1', label: 'BLACK-WHITE (9000)', selectable: true },
        { id: '2', label: 'RED-WHITE (1000)', selectable: true },
      ],
    };

    const expected2 = {
      charRef: { current: null },
      disabled: true,
      highlight: false,
      id: '2',
      key: '2',
      label: 'Size',
      selected: null,
      values: [
        { id: '11', label: '7', selectable: true },
        { id: '12', label: '7.5', selectable: true },
      ],
    };

    expect(renderer).toHaveBeenCalledTimes(2);

    const args1 = renderer.mock.calls[0][0];
    expect(typeof args1.select).toBe('function');
    expect(args1).toMatchObject(expected1);

    const args2 = renderer.mock.calls[1][0];
    expect(typeof args2.select).toBe('function');
    expect(args2).toMatchObject(expected2);
  });
});
