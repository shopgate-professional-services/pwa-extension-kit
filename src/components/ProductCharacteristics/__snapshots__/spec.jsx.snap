// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ProductCharacteristics should render 1`] = `
<Component
  render={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "charRef": Object {
              "current": null,
            },
            "disabled": false,
            "highlight": false,
            "id": "1",
            "key": "1",
            "label": "Color",
            "select": [Function],
            "selected": null,
            "values": Array [
              Object {
                "id": "1",
                "label": "BLACK-WHITE (9000)",
                "selectable": true,
              },
              Object {
                "id": "2",
                "label": "RED-WHITE (1000)",
                "selectable": true,
              },
            ],
          },
        ],
        Array [
          Object {
            "charRef": Object {
              "current": null,
            },
            "disabled": true,
            "highlight": false,
            "id": "2",
            "key": "2",
            "label": "Size",
            "select": [Function],
            "selected": null,
            "values": Array [
              Object {
                "id": "11",
                "label": "7",
                "selectable": true,
              },
              Object {
                "id": "12",
                "label": "7.5",
                "selectable": true,
              },
            ],
          },
        ],
      ],
      "results": undefined,
    }
  }
>
  <WithProductContext
    WrappedComponent={[Function]}
    render={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "charRef": Object {
                "current": null,
              },
              "disabled": false,
              "highlight": false,
              "id": "1",
              "key": "1",
              "label": "Color",
              "select": [Function],
              "selected": null,
              "values": Array [
                Object {
                  "id": "1",
                  "label": "BLACK-WHITE (9000)",
                  "selectable": true,
                },
                Object {
                  "id": "2",
                  "label": "RED-WHITE (1000)",
                  "selectable": true,
                },
              ],
            },
          ],
          Array [
            Object {
              "charRef": Object {
                "current": null,
              },
              "disabled": true,
              "highlight": false,
              "id": "2",
              "key": "2",
              "label": "Size",
              "select": [Function],
              "selected": null,
              "values": Array [
                Object {
                  "id": "11",
                  "label": "7",
                  "selectable": true,
                },
                Object {
                  "id": "12",
                  "label": "7.5",
                  "selectable": true,
                },
              ],
            },
          ],
        ],
        "results": undefined,
      }
    }
  >
    <Theme>
      <Component
        AppBar={[Function]}
        Drawer={[Function]}
        View={[Function]}
        contexts={
          Object {
            "ProductContext": Object {
              "Consumer": [Function],
            },
          }
        }
      >
        <Consumer>
          <Component
            conditioner={
              Conditioner {
                "conditioners": Map {
                  "product-variants" => [Function],
                },
              }
            }
            options={Object {}}
            productId="123"
            variantId="123-45"
          >
            <Component
              productContext={
                Object {
                  "conditioner": Conditioner {
                    "conditioners": Map {
                      "product-variants" => [Function],
                    },
                  },
                  "options": Object {},
                  "productId": "123",
                  "variantId": "123-45",
                }
              }
              render={
                [MockFunction] {
                  "calls": Array [
                    Array [
                      Object {
                        "charRef": Object {
                          "current": null,
                        },
                        "disabled": false,
                        "highlight": false,
                        "id": "1",
                        "key": "1",
                        "label": "Color",
                        "select": [Function],
                        "selected": null,
                        "values": Array [
                          Object {
                            "id": "1",
                            "label": "BLACK-WHITE (9000)",
                            "selectable": true,
                          },
                          Object {
                            "id": "2",
                            "label": "RED-WHITE (1000)",
                            "selectable": true,
                          },
                        ],
                      },
                    ],
                    Array [
                      Object {
                        "charRef": Object {
                          "current": null,
                        },
                        "disabled": true,
                        "highlight": false,
                        "id": "2",
                        "key": "2",
                        "label": "Size",
                        "select": [Function],
                        "selected": null,
                        "values": Array [
                          Object {
                            "id": "11",
                            "label": "7",
                            "selectable": true,
                          },
                          Object {
                            "id": "12",
                            "label": "7.5",
                            "selectable": true,
                          },
                        ],
                      },
                    ],
                  ],
                  "results": undefined,
                }
              }
            >
              <Component
                conditioner={
                  Conditioner {
                    "conditioners": Map {
                      "product-variants" => [Function],
                    },
                  }
                }
                options={Object {}}
                productId="123"
                render={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        Object {
                          "charRef": Object {
                            "current": null,
                          },
                          "disabled": false,
                          "highlight": false,
                          "id": "1",
                          "key": "1",
                          "label": "Color",
                          "select": [Function],
                          "selected": null,
                          "values": Array [
                            Object {
                              "id": "1",
                              "label": "BLACK-WHITE (9000)",
                              "selectable": true,
                            },
                            Object {
                              "id": "2",
                              "label": "RED-WHITE (1000)",
                              "selectable": true,
                            },
                          ],
                        },
                      ],
                      Array [
                        Object {
                          "charRef": Object {
                            "current": null,
                          },
                          "disabled": true,
                          "highlight": false,
                          "id": "2",
                          "key": "2",
                          "label": "Size",
                          "select": [Function],
                          "selected": null,
                          "values": Array [
                            Object {
                              "id": "11",
                              "label": "7",
                              "selectable": true,
                            },
                            Object {
                              "id": "12",
                              "label": "7.5",
                              "selectable": true,
                            },
                          ],
                        },
                      ],
                    ],
                    "results": undefined,
                  }
                }
                variantId="123-45"
              >
                <ProductCharacteristics
                  conditioner={
                    Conditioner {
                      "conditioners": Map {
                        "product-variants" => [Function],
                      },
                    }
                  }
                  finishTimeout={0}
                  navigate={[Function]}
                  options={Object {}}
                  productId="123"
                  render={
                    [MockFunction] {
                      "calls": Array [
                        Array [
                          Object {
                            "charRef": Object {
                              "current": null,
                            },
                            "disabled": false,
                            "highlight": false,
                            "id": "1",
                            "key": "1",
                            "label": "Color",
                            "select": [Function],
                            "selected": null,
                            "values": Array [
                              Object {
                                "id": "1",
                                "label": "BLACK-WHITE (9000)",
                                "selectable": true,
                              },
                              Object {
                                "id": "2",
                                "label": "RED-WHITE (1000)",
                                "selectable": true,
                              },
                            ],
                          },
                        ],
                        Array [
                          Object {
                            "charRef": Object {
                              "current": null,
                            },
                            "disabled": true,
                            "highlight": false,
                            "id": "2",
                            "key": "2",
                            "label": "Size",
                            "select": [Function],
                            "selected": null,
                            "values": Array [
                              Object {
                                "id": "11",
                                "label": "7",
                                "selectable": true,
                              },
                              Object {
                                "id": "12",
                                "label": "7.5",
                                "selectable": true,
                              },
                            ],
                          },
                        ],
                      ],
                      "results": undefined,
                    }
                  }
                  variantId="123-45"
                  variants={
                    Object {
                      "characteristics": Array [
                        Object {
                          "id": "1",
                          "label": "Color",
                          "values": Array [
                            Object {
                              "id": "1",
                              "label": "BLACK-WHITE (9000)",
                            },
                            Object {
                              "id": "2",
                              "label": "RED-WHITE (1000)",
                            },
                          ],
                        },
                        Object {
                          "id": "2",
                          "label": "Size",
                          "values": Array [
                            Object {
                              "id": "11",
                              "label": "7",
                            },
                            Object {
                              "id": "12",
                              "label": "7.5",
                            },
                          ],
                        },
                      ],
                      "products": Array [
                        Object {
                          "characteristics": Object {
                            "1": "1",
                            "2": "11",
                          },
                          "id": "32050590001000",
                        },
                        Object {
                          "characteristics": Object {
                            "1": "2",
                            "2": "12",
                          },
                          "id": "32050590000950",
                        },
                      ],
                    }
                  }
                />
              </Component>
            </Component>
          </Component>
        </Consumer>
      </Component>
    </Theme>
  </WithProductContext>
</Component>
`;
